common/filesystem.h

```diff
@@ -0,0 +1,1374 @@
+enum class OpenAccess {
+  ReadOnly = O_RDONLY,
+  WriteOnly = O_WRONLY,
+  ReadWrite = O_RDWR,
+};
+
+// The different creation styles available when opening a file or directory.
+//
+// Because these are by far the most common parameters and they have unambiguous
+// names, the enumerators are also available directly within the namespace.
+enum class CreationFlags {
+  OpenExisting = 0,
+  OpenAlways = O_CREAT,
+  CreateAlways = O_CREAT | O_TRUNC,
+  CreateNew = O_CREAT | O_EXCL,
+};
+using CreationFlags::CreateAlways;
+using CreationFlags::CreateNew;
+using CreationFlags::OpenAlways;
+using CreationFlags::OpenExisting;
+
+// General flags to control the behavior of opening files that aren't covered by
+// other more specific flags.
+enum class OpenFlags {
+  None = 0,
+  Append = O_APPEND,
+  Truncate = O_TRUNC,
+};
+
+// Flags controlling which permissions should be checked in an `Access` call.
+//
+// These permissions can also be combined with the `|` operator, so `R_OK |
+// W_OK` checks for both read and write access.
+enum class AccessCheck : int {
+  Exists = F_OK,
+  Read = R_OK,
+  Write = W_OK,
+  Execute = X_OK,
+};
+inline auto operator|(AccessCheck lhs, AccessCheck rhs) -> AccessCheck {
+  return static_cast<AccessCheck>(static_cast<int>(lhs) |
+                                  static_cast<int>(rhs));
+}
+
+// The underlying type that should be used to model the mode of a type.
+using ModeType = mode_t;
+
+// Enumeration of the different file types recognized.
+//
+// In addition to the specific type values being arranged for ease of use with
+// the POSIX APIs, the underlying type of the enum is arranged to use the common
+// mode type.
+enum class FileType : ModeType {
+  // Portable file types that need to be supported across platform
+  // implementations.
+  Directory = S_IFDIR,
+  RegularFile = S_IFREG,
+  SymbolicLink = S_IFLNK,
+
+  // Non-portable Unix-variant specific types.
+  FIFO = S_IFIFO,
+  CharDevice = S_IFCHR,
+  BlockDevice = S_IFBLK,
+  Socket = S_IFSOCK,
+
+  // Mask for the Unix-variant types to allow easy extraction.
+  Mask = S_IFMT,
+};
+
+// Forward declarations of various types that appear in APIs.
+class DirRef;
+class Dir;
+class RemovingDir;
+template <OpenAccess A>
+class FileRef;
+template <OpenAccess A>
+class File;
+class FdError;
+class PathError;
+namespace Internal {
+class FileRefBase;
+}  // namespace Internal
+
+// Returns a constant `Dir` object that models the open current working
+// directory.
+//
+// Whatever the working directory of the process is will be used as the base for
+// any relative path operations on this object. For example, on Unix-like
+// systems, `Cwd().Stat("some/path")` is equivalent to `stat("some/path")`.
+consteval auto Cwd() -> Dir;
+
+// Creates a temporary directory and returns a removing directory handle to it.
+//
+// Each directory created should be unique and newly created by the call. It is
```

Comment:
```suggestion
// Each directory created will be unique and newly created by the call. It is
```
I believe this can be phrased as a promise of the API, rather than a condition to be met (perhaps by the caller)